integration_tests/run.py

```diff
@@ -0,0 +1,239 @@
+
+
+class TestContextManager(object):
+  """Manages all resources for a given run of tests.
+
+  Responsible for setting up tests (i.e. creating a unique dataset) and
+  cleaning up all resources after tests have been completed.
+  """
+
+  def __init__(self, known_args):
+    self.staging_location = known_args.staging_location
+    self.temp_location = known_args.temp_location
+    self.logging_location = known_args.logging_location
+    self.project = known_args.project
+
+  def __enter__(self):
+    self.credentials = GoogleCredentials.get_application_default()
+    client = bigquery.Client(project=self.project)
+    self.dataset_id = 'integration_tests_{}'.format(
+        datetime.now().strftime('%Y%m%d_%H%M%S'))
+    dataset_ref = client.dataset(self.dataset_id)
+    self.dataset = bigquery.Dataset(dataset_ref)
+    _ = client.create_dataset(self.dataset)
+    return self
+
+  def __exit__(self, *args):
+    client = bigquery.Client(project=self.project)
+    tables = client.list_dataset_tables(self.dataset)
+    for table in tables:
+      client.delete_table(table)
+    client.delete_dataset(self.dataset)
+
+
+def main():
+  tests_json = []
+  for filename in _get_test_files():
+    tests_json.append(_load_test_json(filename))
+
+  with TestContextManager(_get_known_args()) as context:
+    tests = _create_tests(tests_json, context)
+    pool = multiprocessing.Pool(processes=len(tests))
+    results = []
+    for test in tests:
+      results.append(pool.apply_async(func=_run_test, args=(test, context)))
+    pool.close()
+    pool.join()
+
+  _print_errors(results, tests)
+
+
+def _get_known_args():
+  parser = argparse.ArgumentParser()
+  parser.add_argument('--project', required=True)
+  parser.add_argument('--staging_location', required=True)
+  parser.add_argument('--temp_location', required=True)
+  parser.add_argument('--logging_location', required=True)
+  known_args, _ = parser.parse_known_args()
+  return known_args
+
+
+def _get_test_files():
+  """Gets all json files in integration_tests directory and subdirectories."""
+  test_file_path = os.path.join(os.getcwd(), 'integration_tests')
+  for root, _, files in os.walk(test_file_path):
+    for filename in files:
+      if filename.endswith('.json'):
+        yield os.path.join(root, filename)
+
+
+def _load_test_json(filename):
+  """Loads an integration test JSON object from a file."""
+  with open(filename, 'r') as f:
+    test = json.loads(f.read())
+    _validate_test(test, filename)
+    return test
+
+
+def _validate_test(test, filename):
+  required_keys = ['test_name', 'table_name', 'input_pattern']
+  for key in required_keys:
+    if key not in test:
+      raise ValueError('Test case in {} is missing required key: {}'.format(
+          filename, key))
+
+
+def _create_tests(tests_json, context):
+  """Creates TestCases from test JSON objects."""
+  return [TestCase(context, **test_json) for test_json in tests_json]
+
+
+def _run_test(test, context):
```

Comment:
Can this be within TestCase class? I think it's more natural to have it in there.

Reply:
Done.

Reply:
Unfortunately we can't do this because instance methods apparently cannot be pickled.